content/blog/2019-09-25-making-function-calls-work.md

```diff
@@ -0,0 +1,299 @@
+
+To pass a return value back to the caller, we use the following approach. When we handle a `ret` operation, if there is a return value, we add a special variable `_ ret` to the environment and map it to the return value. Note that this will never collide with any existing variables as there is space in the name. Then, if the caller is expecting a return value (which it will in instructions of the form `v: type = call func_name`), it can check for the existence of `_ ret` in the callee's environment. We do some basic type-checking here, comparing return types and variable types, to make the Bril programmer's life easier.
+
+Below is an example program that demonstrates this functionality.
+
+    # Once again, we print 100.
+    main {
+      v0: int = call get_hundred;
+      print v0;
+    }
+    
+    get_hundred: int {  # The ": int" is required.
+      v0: int = const 100;
+      ret v0;
+    }
+
+Since we now have the ability to pass arguments to functions and get return values, we can now write some interesting Bril programs. Below is a Bril program that prints the 10th Fibonacci number.
+
+    # Print fib(10)
+    main {
+        v0: int = const 10;
+        fib10: int = call fib v0;
+        print fib10;
+    }
+
+    # Return true if n <= 1, false otherwise.
+    lte_one n: bool {
+        one: int = const 1;
+        lto: bool = le n one;
+        ret lto;
+    }
+
+    # Return fib(10).
+    fib n: int {
+        base: bool = call lte_one n;
+        br base return continue;
+    return:
+        ret n;
+    continue:
+        one: int = const 1;
+        prev: int = sub n one;
+        prev2: int = sub prev one;
+        fib1: int = call fib prev;
+        fib2: int = call fib prev2;
+        ans: int = add fib1 fib2;
+        ret ans;
+    }
+
+**Note**: To call a function on the right-hand side of an assignment, the function must declare a return type. We choose to enforce this to improve readability. The following program will fail with the error message `function func does not return`.
+
+    # Fails!
+    main {
+      v0: int = call func;
+      print v0;
+    }
+
+    func {
+      v0: int = const 100;
+      ret v0;
+    }
+
+The opposite is allowed: functions that have a declared return type *may* be called as standalone instructions. The following program succeeds and prints 100.
+
+    # Print 100
+    main {
+      call func;
+    }
+
+    func: int {
+      v0: int = const 100;
+      print v0;
+      ret v0;
+    }
+
+
+#### Optional type annotations for function parameters
+Function definitions are now more complex, and with this added syntactic complexity comes a loss of readability. To fix this, we introduce the notion of type annotations for function parameters. Since we do not do support static type checking, these type annotations serve no operational purpose. As a result, all we have to do is update the language grammar. Below is an updated version of the above program with type annotations.
```

Comment:
Got it. So maybe the answer to my question above is "the interpreter doesn't care about the types, so we left them off initially." That's fine, but I do think the "full version" of the system is probably the fully-typed version.